src/math/src/number.rs

```diff
@@ -1,18 +1,523 @@
+        todo!()
+    }
+
+    /// Calculate factorial of a given number
+    /// The number is not limited to integer, it can be a fraction
+    ///
+    /// # Error
+    /// return Error::FactorialNegative if the number is less than 0
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::Zero.factorial(), Number::One);
+    /// assert_eq!(Number::from(5).factorial(), Ok(Number::from(120)));
+    /// assert_eq!(Number::from(3.2).factorial().unwrap().to_string(6), "7.756689");
+    /// assert!(Number::from(-1).factorial().is_err());
+    /// ```
+    pub fn factorial(&self) -> Result<Self> {
+        todo!()
+    }
+
+    /// Returns the logarithm of the number with respect to an arbitrary `base`.
+    ///
+    /// # Error
+    /// Error::LogUndefinedBase if the `base` is less or equal than 0
+    /// Error::LogUndefinedNumber if the number is less or equal than 0
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// # fn main() -> math::Result<()> {
+    ///     assert!(Number::random().log(0).is_err());
+    ///     assert!(Number::random().log(-1.2).is_err());
+    ///     assert!(Number::ZERO.log(Number::random()).is_err());
+    ///     assert!(Number::from(-0.3).log(Number::random()).is_err());
+    ///
+    ///     let a = Number::random();
+    ///     let base = Number::random();
+    ///     let b = Number::from(2);
+    ///
+    ///     // Number same as base
+    ///     assert_eq!(a.log(a)?, Number::ONE);
+    ///     // Product rule log(xy) == log(x) + log(y)
+    ///     assert_eq!(a.mul(b)?.log(base), a.log(base)?.add(b.log(base)?));
+    ///     // Quotient rule log(x/y) == log(x) - log(y)
+    ///     assert_eq!(a.div(b)?.log(base), a.log(base)?.sub(b.log(base)?));
+    ///     // Log of power log(x^y) == y * log(x)
+    ///     assert_eq!(a.pow(b)?.log(base), b.mul(a.log(base)?));
+    ///     // Log of one
+    ///     assert_eq!(Number::ONE.log(base)?, Number::ZERO);
+    ///     // Log reciprocal log(1/x) = -ln(x);
+    ///     assert_eq!(Number::ONE.div(a)?.log(base), a.log(base)?.mul(-1));
+    ///     # Ok(())
+    /// # }
+    /// ```
+    pub fn log(&self, base: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Same as `Number::log` with `base` of 2
+    pub fn log2(&self) -> Result<Self> {
+        todo!()
+    }
+
+    /// Same as `Number::log` with `base` of `Number::E`
+    pub fn ln(&self) -> Result<Self> {
+        todo!()
+    }
+
+    /// Same as `Number::log` with `base` of 10
+    pub fn log10(&self) -> Result<Self> {
+        todo!()
+    }
+
+    /// Returns the nth root of a number
+    ///
+    /// # Error
+    /// Error::ZeroNthRoot if the `nth` is 0
+    /// Error::NegativeRoot if the number is negative
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// # fn main() -> math::Result<()> {
+    ///     assert!(Number::random().root(0).is_err());
+    ///     assert!(Number::from(-1).root(Number::random()).is_err());
```

Comment:
This should only fail when the root `n` is even.

Reply:
Fixed, thanks!